pkg/pillar/cmd/zedrouter/flowstats.go

```diff
@@ -670,9 +740,114 @@ func DNSStopMonitor(bnNum int) {
+	items := pub.GetAll()
+	for _, st := range items {
+		netstatus = st.(types.NetworkInstanceStatus)
+		if netstatus.Type != types.NetworkInstanceTypeSwitch || netstatus.BridgeNum != bnNum {
+			continue
+		}
+		vifInfo = netstatus.Vifs
+		break
+	}
+	if len(vifInfo) == 0 { // there is no Mac on the bridge
+		log.Tracef("checkDADProbe: no mac on the bridge")
+		return
+	}
+
+	var etherPkt *layers.Ethernet
+	etherLayer := packet.Layer(layers.LayerTypeEthernet)
+	if etherLayer != nil {
+		etherPkt, _ = etherLayer.(*layers.Ethernet)
+		for _, vif := range vifInfo {
+			if strings.Compare(etherPkt.SrcMAC.String(), vif.MacAddr) == 0 {
+				foundSrcMac = true
+				break
+			}
+		}
+	}
+	if !foundSrcMac {
+		log.Tracef("checkDADProbe: pkt no dst mac for us\n")
+		return
+	}
+
+	ip6Layer := packet.Layer(layers.LayerTypeIPv6)
+	if ip6Layer == nil {
+		return
+	}
+	ip6 := ip6Layer.(*layers.IPv6)
+	// We are looking for ICMPv6 Neighbor solicitation packet that
+	// tries to find if the address calculated locally is a duplicate.
+	// Such packets have a source IP of all zeroes (::)
+	if ip6.SrcIP.String() != "::" {
+		return
+	}
+
+	icmp6Layer := packet.Layer(layers.LayerTypeICMPv6NeighborSolicitation)
+	if icmp6Layer == nil {
+		return
+	}
+	for _, vif := range vifInfo {
+		if strings.Compare(vif.MacAddr, etherPkt.SrcMAC.String()) == 0 {
+			icmp6, _ := icmp6Layer.(*layers.ICMPv6NeighborSolicitation)
+			log.Tracef("ICMPv6: TargetAddress %s", icmp6.TargetAddress.String())
+
+			if _, ok := netstatus.IPAssignments[vif.MacAddr]; !ok {
+				log.Functionf("checkDADProbe: mac %v assign new IPv6 address %v\n",
+					vif.MacAddr, icmp6.TargetAddress)
+				addrs := types.AssignedAddrs{IPv6Addrs: []net.IP{icmp6.TargetAddress}}
+				netstatus.IPAssignments[vif.MacAddr] = addrs
+				needUpdate = true
+			} else {
+				if !isAddrPresent(netstatus.IPAssignments[vif.MacAddr].IPv6Addrs, icmp6.TargetAddress) {
+					log.Functionf("checkDADProbe: update mac list %v, new IPv6 %v\n",
+						vif.MacAddr, icmp6.TargetAddress)
+					addrs := netstatus.IPAssignments[vif.MacAddr]
+					addrs.IPv6Addrs = append(addrs.IPv6Addrs, icmp6.TargetAddress)
+					netstatus.IPAssignments[vif.MacAddr] = addrs
+					needUpdate = true
+				}
+			}
+			ipv4Addr, snoopedIPv6s, _ := lookupVifIPTrig(ctx, vif.MacAddr)
+			vifTrig.MacAddr = vif.MacAddr
+			vifTrig.IPv4Addr = ipv4Addr
+			if !isAddrPresent(snoopedIPv6s, icmp6.TargetAddress) {
+				vifTrig.IPv6Addrs = append(snoopedIPv6s, icmp6.TargetAddress)
+			} else {
+				vifTrig.IPv6Addrs = snoopedIPv6s
+			}
+		}
+	}
+
+	if needUpdate {
+		log.Functionf("checkDADProbe: need update %v, %v\n", vifInfo, netstatus.IPAssignments)
+		pub := ctx.pubNetworkInstanceStatus
+		pub.Publish(netstatus.Key(), netstatus)
+		ctx.pubAppVifIPTrig.Publish(vifTrig.MacAddr, vifTrig)
+		checkAndPublishDhcpLeases(ctx)
+	}
+}
+
+func isAddrPresent(list []net.IP, addr net.IP) bool {
+	for i := 0; i < len(list); i++ {
+		if list[i].String() == addr.String() {
```

Comment:
Here you could use IP.Equal and avoid conversion to string

Reply:
Changed the check to IP.Equal